zcash_client_backend/src/data_api/chain.rs

```diff
@@ -213,6 +213,168 @@ pub trait BlockSource {
+///         Ok(highest_block.map(|&block| BlockHeight::from_u32(block.height as u32)))
+///     }
+///
+///     async fn insert(&mut self, mut compact_blocks: Vec<CompactBlock>) -> Result<(), Self::Error> {
+///         self.cached_blocks.append(&mut compact_blocks);
+///         Ok(())
+///     }
+///
+///     fn truncate(&mut self, block_height: BlockHeight) -> Result<(), Self::Error> {
+///         self.cached_blocks
+///             .retain(|block| block.height <= block_height.into());
+///         Ok(())
+///     }
+///
+///     fn mark_as_scanned(&mut self, _range: &ScanRange) -> Result<(), Self::Error> {
+///         todo!()
+///     }
+///
+///     async fn delete_scanned(&mut self) -> Result<(), Self::Error> {
+///         todo!()
+///     }
+/// }
+///
+/// // Example usage
+/// let mut block_cache = ExampleBlockCache {
+///     cached_blocks: Vec::new(),
+/// };
+/// let range = ScanRange::from_parts(
+///     BlockHeight::from_u32(1)..BlockHeight::from_u32(3),
+///     ScanPriority::Historic,
+/// );
+/// let account = zip32::AccountId::ZERO;
+/// let extsk = sapling::zip32::ExtendedSpendingKey::master(&[]);
+/// let dfvk = extsk.to_diversifiable_full_viewing_key();
+///
+/// let compact_block1 = zcash_client_backend::scanning::testing::fake_compact_block(
+///     1u32.into(),
+///     zcash_primitives::block::BlockHash([0; 32]),
+///     sapling::Nullifier([0; 32]),
+///     &dfvk,
+///     zcash_primitives::transaction::components::amount::NonNegativeAmount::const_from_u64(5),
+///     false,
+///     None,
+/// );
+/// let compact_block2 = zcash_client_backend::scanning::testing::fake_compact_block(
+///     2u32.into(),
+///     zcash_primitives::block::BlockHash([0; 32]),
+///     sapling::Nullifier([0; 32]),
+///     &dfvk,
+///     zcash_primitives::transaction::components::amount::NonNegativeAmount::const_from_u64(5),
+///     false,
+///     None,
+/// );
+/// let compact_blocks = vec![compact_block1, compact_block2];
+///
+/// // Insert into the block cache
+/// let mut rt = tokio::runtime::Runtime::new().unwrap();
+/// rt.block_on(block_cache.insert(compact_blocks.clone()));
+/// assert_eq!(block_cache.cached_blocks.len(), 2);
+///
+/// // Find highest block in the block cache
+/// let cache_tip = block_cache.cache_tip(None).unwrap();
+/// assert_eq!(cache_tip, Some(BlockHeight::from_u32(2)));
+///
+/// // Read from the block cache
+/// let blocks_from_cache = block_cache.read(&range).unwrap();
+/// assert_eq!(blocks_from_cache, compact_blocks);
+///
+/// // Truncate the block cache
+/// block_cache.truncate(BlockHeight::from_u32(1)).unwrap();
+/// assert_eq!(block_cache.cached_blocks.len(), 1);
+/// assert_eq!(
+///     block_cache.cache_tip(None).unwrap(),
+///     Some(BlockHeight::from_u32(1))
+/// );
+/// ```
+pub trait BlockCache: BlockSource + Send + Sync {
+    /// Returns a range of compact blocks from the cache.
+    fn read(&self, range: &ScanRange) -> Result<Vec<CompactBlock>, Self::Error>;
```

Comment:
What are the semantics  that are required if not all blocks in the scan range exist in the cache? Are partial reads allowed, or does that need to be represented as an error?

Reply:
I imagined that the error types and conditions would be decided on implementation so it is flexible for the consumer

Reply:
On second thought partial reads probably should be allowed in the case that a block cache is storing compact blocks on disk and therefore can resume if the whole scan range had not be stored yet instead of downloading all the blocks in the scan range. This would be a nice addition to the sync implementation but would be an optimisation as oppose to a key component and im not sure how often this would occur. As other wallets (i.e. probably zingo) will store the block cache in memory for performance, this will not apply in those implemenations. The other option would be to define an additional block cache error enum where one of the variants is associated with not reading all blocks in scan range so the sync implementation can mathc on this error instead of checking the Vec<CompactBlock> matches the size of the scan range

Reply:
Isn't the availability of blocks in the cache orthogonal to the backing store?  If the cache is in-memory it still must go through a non-deterministic population process?  e.g. downloading the blocks from a chain-attached server? Apologies if I am misunderstanding something fundamental..   perhaps that concern is handled elsewhere?

Reply:
Okay. Please document this decision in the method documentation.

Reply:
It seems reasonable to me to allow short reads (but not reads with gaps). It is fairly easy for consumers to handle that, and it still fits the current `Result<Vec<CompactBlock>, Self::Error>` return type.